### PR TITLE
[Feature] Add cross-database migration, gen_job and migration built-in subagents

### PR DESCRIPTION
**Pull request by stukid**
## Why

Datus needs cross-database migration capability (DuckDB → Greenplum/StarRocks) with automatic type mapping, data transfer, and post-migration reconciliation. The existing ETL workflow was config-based only and lacked cross-database support.

Related: #563 (follow-up for adapter-layer identifier quoting)

### Core Tool Layer
- `execute_ddl` adds `database` parameter for multi-connector routing
- New `transfer_query_result` tool for cross-database data movement (batch INSERT via `execute_insert`, NaT/NaN → None conversion, 1M row limit)
- `_get_connector` routing fix: uses `db_name` as both namespace and logic_name for cross-database access
- `list_databases` returns `available` status and `type` info in multi-connector mode
- `describe_table` avoids passing logical name to connector's `get_schema` in multi-connector mode
- SQLAlchemy transaction commit after DDL/transfer to prevent table locks
- Dialect-aware column quoting: backticks for MySQL/StarRocks/Hive/Spark/BigQuery/ClickHouse, double quotes for PostgreSQL/Greenplum/Snowflake/Trino

### Migration Helpers (`datus/tools/migration/`)
- `type_mapping.py`: DuckDB → Greenplum/StarRocks type mapping with unsupported type detection
- `target_profiles.py`: Greenplum/StarRocks DDL generation with automatic key selection
- `reconciliation.py`: 7-item reconciliation check SQL generation

### Built-in Subagents
- **gen_job** (single-database ETL): DDL + DML, skills: gen-table + table-validation, max_turns=30
- **migration** (cross-database): DDL + DML + transfer_query_result, skills: data-migration + table-validation, max_turns=40
- Both registered across all 8 registration points (node_type, node factory, sub_agent_task_tool, constants)
- `GenSQLAgenticNode._needs_multi_connector()` auto-detects when to use multi-connector mode

### Safety Guards
- `transfer_query_result` error messages explicitly say "Do NOT fall back" when connector unavailable
- Migration prompt and skill hard-ban database fallback
- `list_databases` connectivity check filters unavailable databases

### Documentation
- Design doc: `docs/cross-db-migration-design.zh.md`
- User docs: gen_table, gen_job, migration (EN + ZH)
- Updated builtin_subagents overview (EN + ZH)
- mkdocs navigation updated

### Unit Tests (221 tests)
- `tests/unit_tests/tools/migration/test_type_mapping.py` — 77 tests: type mapping, unsupported types, aliases, case insensitivity
- `tests/unit_tests/tools/migration/test_target_profiles.py` — 28 tests: DDL generation, key selection, StarRocks profiles
- `tests/unit_tests/tools/migration/test_reconciliation.py` — 13 tests: 7 check types, edge cases
- `tests/unit_tests/tools/func_tool/test_database.py` — added: execute_ddl routing, transfer_query_result (replace/append/failure/limits), connector routing, fallback
- `tests/unit_tests/agent/node/test_gen_job_agentic_node.py` — 15 tests: init, tools, factory
- `tests/unit_tests/agent/node/test_migration_agentic_node.py` — 15 tests: init, tools, factory, execute_stream

### Integration Tests (14 tests, require Docker)
- `tests/integration/tools/test_migration_integration.py` — DuckDB → Greenplum (3) + DuckDB → StarRocks (3) + end-to-end (2)
- `tests/integration/tools/test_migration_subagent.py` — MigrationAgenticNode tool chain: list_databases → describe_table → execute_ddl → transfer → verify (6)

```bash
# Unit tests
uv run pytest tests/unit_tests/tools/migration/ tests/unit_tests/tools/func_tool/test_database.py tests/unit_tests/agent/node/test_gen_job_agentic_node.py tests/unit_tests/agent/node/test_migration_agentic_node.py -v

# Integration tests (requires Docker)
cd datus-db-adapters/datus-greenplum && docker compose up -d
cd datus-db-adapters/datus-starrocks && docker compose up -d
uv run pytest tests/integration/tools/test_migration_subagent.py tests/integration/tools/test_migration_integration.py -v --timeout=30
```

## Summary by CodeRabbit

* **New Features**
  * Added gen_job (single-db ETL) and migration (cross-db migration) subagents, new ETL/migration tools (write & transfer), scoped-table exploration mode, filesystem tooling, and OpenAI prompt caching.
* **Bug Fixes**
  * Improved DB connector routing, safer SQL file handling, and JSON/SQL sanitization for model outputs.
* **Documentation**
  * Added extensive docs and examples for new subagents, skills, prompts, and migration/DDL/type-mapping guidance.

### CODE REVIEW

File: datus/tools/func_tool/database.py [hunk 1]
```diff
@@ -216,8 +216,19 @@ def _get_connector(self, database: Optional[str] = None) -> BaseSqlConnector:
...
-        connector = self._db_manager.get_conn(self._namespace, db_name)
+        # Fetch from db_manager.
+        # Each database in service.databases is its own namespace (see AgentConfig.namespaces).
+        # For cross-database scenarios, use db_name as both namespace and logic_name.
+        try:
+            connector = self._db_manager.get_conn(db_name, db_name)
+        except (KeyError, ValueError) as e:
+            # Fallback to current namespace for backward compatibility
+            logger.debug(f"Falling back to namespace lookup for '{db_name}': {e}")
+            connector = self._db_manager.get_conn(self._namespace, db_name)
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:234]: Fixed in 0f27f550. Added source_sql validation (single-statement, read-only), strict connector lookup for explicit database param, and moved TRUNCATE before empty check.

File: docs/subagent/gen_table.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+| Feature | gen_table | gen_job |
+|---------|-----------|---------|
+| Table creation (CTAS) | Yes | Yes |
+| Table creation (DDL) | Yes | Yes |
+| Data writes (INSERT/UPDATE/DELETE) | No | Yes |
+| Cross-database migration | No | Yes |
+| User confirmation (ask_user) | Always required | Available |
+| Default max_turns | 20 | 30 |
+
+Use **gen_table** when you need interactive table creation with user confirmation. Use **gen_job** when you need ETL pipelines or cross-database migration.
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/subagent/gen_table.md]: Fixed in 4dfc6dfc.

File: datus/agent/node/agentic_node.py [hunk 3]
```diff
@@ -750,6 +750,8 @@ def _get_tool_category(self, tool_name: str) -> str:
         if tool_name.startswith("db_") or tool_name in [
             "list_tables",
             "describe_table",
+            "execute_ddl",
+            "execute_write",
             "execute_sql",
             "get_sample_data",
         ]:
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/agent/node/agentic_node.py:758]: Fixed in 4dfc6dfc.

File: datus/tools/func_tool/database.py [hunk 4]
```diff
@@ -1123,22 +1179,350 @@ def execute_ddl(self, sql: str) -> FuncToolResult:
...
+                return FuncToolResult(
+                    success=0,
+                    error="Source database connector does not support pandas execution.",
+                )
+            source_result = source_conn.execute_pandas(source_sql)
+            if not source_result.success:
+                return FuncToolResult(success=0, error=f"Source query failed: {source_result.error}")
+            df = source_result.sql_return
+        except Exception as e:
+            return FuncToolResult(success=0, error=f"Source query execution failed: {str(e)}")
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:1452]: Fixed in 0f27f550. Added source_sql validation (single-statement, read-only), strict connector lookup for explicit database param, and moved TRUNCATE before empty check.

File: docs/subagent/gen_table.zh.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+| 能力 | gen_table | gen_job |
+|------|-----------|---------|
+| 建表 (CTAS) | 支持 | 支持 |
+| 建表 (DDL) | 支持 | 支持 |
+| 数据写入 (INSERT/UPDATE/DELETE) | 不支持 | 支持 |
+| 跨库迁移 | 不支持 | 支持 |
+| 用户确认 (ask_user) | 必须 | 可选 |
+| 默认 max_turns | 20 | 30 |
+
+需要交互式建表时用 **gen_table**，需要 ETL 流水线或跨库迁移时用 **gen_job**。
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/subagent/gen_table.zh.md]: Fixed in 4dfc6dfc.

File: datus/tools/migration/reconciliation.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    if distinct_cols:
+        src_parts = [f"COUNT(DISTINCT {c}) AS {c}_distinct" for c in distinct_cols]
+        tgt_parts = [f"COUNT(DISTINCT {c}) AS {c}_distinct" for c in distinct_cols]
+        checks.append(
+            {
+                "name": "distinct_count",
+                "source_query": f"SELECT {', '.join(src_parts)} FROM {source_table}",
+                "target_query": f"SELECT {', '.join(tgt_parts)} FROM {target_table}",
+            }
+        )
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/migration/reconciliation.py]: Fixed in 4dfc6dfc.

File: datus/tools/func_tool/database.py [hunk 7]
```diff
@@ -1123,22 +1179,350 @@ def execute_ddl(self, sql: str) -> FuncToolResult:
...
+        if mode == "replace":
+            try:
+                truncate_result = target_conn.execute_ddl(f"TRUNCATE TABLE {target_table}")
+                if not truncate_result.success:
+                    return FuncToolResult(
+                        success=0,
+                        error=f"Failed to truncate target table: {truncate_result.error}",
+                    )
+            except Exception as e:
+                return FuncToolResult(success=0, error=f"Failed to truncate target table: {str(e)}")
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py]: Fixed in 0f27f550. Added source_sql validation (single-statement, read-only), strict connector lookup for explicit database param, and moved TRUNCATE before empty check.

File: datus/tools/func_tool/database.py [hunk 8]
```diff
@@ -1123,22 +1179,346 @@ def execute_ddl(self, sql: str) -> FuncToolResult:
...
+                    "message": "Write executed successfully",
+                    "sql": normalized_sql,
+                    "sql_type": sql_type.value,
+                    "row_count": row_count,
+                    "database": database or self._default_database,
+                    "dry_run": dry_run,
+                }
+            )
+        except Exception as e:
+            return FuncToolResult(success=0, error=f"Write execution failed: {str(e)}")
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:1351]: Fixed in 4dfc6dfc.

File: datus/tools/func_tool/sub_agent_task_tool.py [hunk 9]
```diff
@@ -331,6 +348,20 @@ def _create_builtin_node(self, subagent_type: str):
...
+                agent_config=self.agent_config,
+                execution_mode="interactive",
+            )
+        elif subagent_type == "migration":
+            from datus.agent.node.migration_agentic_node import MigrationAgenticNode
+
+            return MigrationAgenticNode(
+                agent_config=self.agent_config,
+                execution_mode="interactive",
+            )
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/sub_agent_task_tool.py:383]: Fixed in 4dfc6dfc.

File: datus/tools/migration/reconciliation.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    checks = []
+    has_keys = bool(key_columns)
+
+    # 1. Row count
+    checks.append(
+        {
+            "name": "row_count",
+            "source_query": f"SELECT COUNT(*) AS row_count FROM {source_table}",
+            "target_query": f"SELECT COUNT(*) AS row_count FROM {target_table}",
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/migration/reconciliation.py:72]: Fixed in 4dfc6dfc.

File: datus/api/models/cli_models.py [hunk 11]
```diff
@@ -460,6 +460,14 @@ class UserInteractionInput(BaseModel):
         description="List of user answers, one per request. Single-select: ['key'], multi-select: ['key1', 'key2'].",
     )
 
+    @field_validator("input", mode="before")
+    @classmethod
+    def normalize_input(cls, v):
+        """Accept legacy List[str] format and normalize to List[List[str]]."""
+        if isinstance(v, list) and v and isinstance(v[0], str):
+            return [[item] for item in v]
+        return v
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/api/models/cli_models.py:282]: The `normalize_input` validator is pre-existing code, not modified in this PR.

File: datus/agent/node/gen_table_agentic_node.py [hunk 12]
```diff
@@ -102,6 +106,16 @@ def _setup_db_tools(self):
...
+    def _setup_filesystem_tools(self):
+        """Setup filesystem tools."""
+        try:
+            root_path = self._resolve_workspace_root()
+            self.filesystem_func_tool = FilesystemFuncTool(root_path=root_path)
+            self.tools.extend(self.filesystem_func_tool.available_tools())
+            logger.debug(f"Setup filesystem tools with root path: {root_path}")
+        except Exception as e:
+            logger.error(f"Failed to setup filesystem tools: {e}")
+
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/agent/node/gen_table_agentic_node.py:123]: Pre-existing pattern. gen_table_agentic_node's filesystem tool init has been logging and continuing since before this PR. Will evaluate separately.

File: datus/tools/func_tool/database.py [hunk 13]
```diff
@@ -1123,22 +1179,346 @@ def execute_ddl(self, sql: str) -> FuncToolResult:
...
+                            values.append("TRUE" if val else "FALSE")
+                        elif isinstance(val, (int, float)):
+                            values.append(str(val))
+                        else:
+                            escaped = str(val).replace("'", "''")
+                            values.append(f"'{escaped}'")
+                    value_rows.append(f"({', '.join(values)})")
+
+                insert_sql = f"INSERT INTO {target_table} ({col_names}) VALUES {', '.join(value_rows)}"
+                result = target_conn.execute_insert(insert_sql)
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:1530]: Acknowledged. The SQL value escaping limitation is tracked in issue #563. The proper fix requires `connector.quote_identifier()` from the adapter layer, which is implemented in datus-db-adapters but not yet released to PyPI.

File: datus/models/openai_compatible.py [hunk 14]
```diff
@@ -158,6 +160,57 @@ def _get_base_url(self) -> Optional[str]:
...
+            [
+                "openai-prompt-cache",
+                self.model_name,
+                agent_name,
+                node_name,
+                namespace,
+                database,
+            ]
+        )
+        return hashlib.sha256(raw_key.encode("utf-8")).hexdigest()[:32]
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/models/openai_compatible.py:212]: The agent_name cache key issue in openai_compatible.py is pre-existing, not introduced by this PR.

File: datus/tools/db_tools/duckdb_connector.py [hunk 15]
```diff
@@ -617,9 +617,32 @@ def get_schema(
...
+                result = self.connection.execute(sql)
+            except duckdb.CatalogException:
+                # In common single-file DuckDB usage, callers may pass the
+                # connector's logical/current database name even though PRAGMA
+                # resolution only needs "schema.table". Retry without the
+                # database qualifier before surfacing the error.
+                if database_name:
+                    fallback_full_name = self.full_name(schema_name=schema_name, table_name=table_name)
+                    fallback_sql = f"PRAGMA table_info('{fallback_full_name}')"
+                    logger.warning(
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/db_tools/duckdb_connector.py:633]: The DuckDB PRAGMA fallback identifier escaping is pre-existing code, not modified in this PR.

File: datus/agent/node/explore_agentic_node.py [hunk 16]
```diff
@@ -215,6 +240,11 @@ async def execute_stream(
 
         user_input = self.input
 
+        # Dynamic scoped context is carried on ExploreNodeInput, so rebuild
+        # tools after input is set to ensure DBFuncTool receives the per-run
+        # table allowlist instead of only static agent.yml configuration.
+        self.setup_tools()
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/agent/node/explore_agentic_node.py:244]: Not introduced by this PR — the explore_agentic_node setup_tools() pattern is pre-existing. Will track separately.

File: tests/integration/tools/test_migration_subagent.py [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+        assert "score" in col_names
+
+    def test_read_source_row_count(self, migration_node):
+        """read_query on source should return 3 rows."""
+        result = migration_node.db_func_tool.read_query(
+            sql="SELECT COUNT(*) AS cnt FROM test_users", database="source_duckdb"
+        )
+        assert result.success == 1
+        # H-7: verify actual count value, not just success
+        assert "3" in str(result.result), f"Expected count=3, got: {result.result}"
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/integration/tools/test_migration_subagent.py:155]: Acknowledged — the string-based assertion is fragile. Will strengthen in a follow-up.

File: docs/cross-db-migration-design.zh.md [hunk 18]
```diff
@@ -0,0 +1,732 @@
...
+
+路由链路：
+
+```
+database="greenplum" → _get_connector("greenplum")
+  → DBManager.get_conn("greenplum", "greenplum")
+  → 查 agent.yml: service.databases.greenplum → DbConfig
+  → ConnectorRegistry 创建 GreenplumConnector
+  → 连接到实际的 test 库
+```
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/cross-db-migration-design.zh.md]: This file has been removed from the PR (it's a local design doc, not intended for merge).

File: tests/unit_tests/agent/node/test_gen_job_agentic_node.py [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+# Licensed under the Apache License, Version 2.0.
+# See http://www.apache.org/licenses/LICENSE-2.0 for details.
+
+"""
+Unit tests for GenJobAgenticNode.
+
+Tests cover:
+- Node creation in workflow and interactive modes
+- Tools setup (DBFuncTool + execute_ddl + execute_write + transfer_query_result)
+- Max turns configuration
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/unit_tests/agent/node/test_gen_job_agentic_node.py:11]: Fixed — test file was rewritten with shared helpers in 4dfc6dfc, header comment is now accurate.

File: tests/integration/tools/test_migration_subagent.py [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+        )
+        assert result.success == 1, f"Transfer failed: {result.error}"
+        assert result.result["rows_transferred"] == 3
+        assert result.result["mode"] == "replace"
+
+        # Verify on target
+        verify = migration_node.db_func_tool.read_query(
+            sql=f"SELECT COUNT(*) AS cnt FROM {TARGET_TABLE}", database="greenplum"
+        )
+        assert verify.success == 1
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/integration/tools/test_migration_subagent.py:205]: Acknowledged — will add explicit row count assertion in a follow-up.

File: tests/unit_tests/agent/node/test_migration_agentic_node.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+            node_id="test_migration",
+            description="Test migration factory",
+            node_type=NodeType.TYPE_MIGRATION,
+            input_data=input_data,
+            agent_config=real_agent_config,
+            tools=[],
+        )
+        assert isinstance(node, MigrationAgenticNode)
+        assert node.input is not None
+        assert node.input.user_message == "Migrate users from duckdb to greenplum"
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/unit_tests/agent/node/test_migration_agentic_node.py]: Both gen_job and migration node test files have been rewritten with shared test helpers. Nightly tests will be added in a follow-up.

File: docs/subagent/gen_table.zh.md [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+
+```
+1. 解析输入 SQL，识别源表和列
+2. 调用 describe_table 获取源表类型信息
+3. 可选调用 read_query LIMIT 验证输出
+4. 生成 CTAS DDL：CREATE TABLE schema.table AS (SELECT ...)
+5. 通过 ask_user 展示完整 DDL → [执行 / 修改 / 取消]
+6. 执行 DDL 并验证（行数 + schema 检查）
+7. 输出摘要
+```
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/subagent/gen_table.zh.md:54]: Will add language identifiers.

File: docs/subagent/migration.md [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+
+## Workflow
+
+```
+Phase 1: Discover databases (list_databases) → inspect source schema
+Phase 2: Generate target DDL with type mapping → create target table
+Phase 3: Transfer data (transfer_query_result)
+Phase 4: Reconcile source vs target (7 mandatory checks)
+Phase 5: Output migration report
+```
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/subagent/migration.md:67]: Will add language tag.

File: docs/design/de-skills-design-summary.md [hunk 24]
```diff
@@ -0,0 +1,305 @@
...
+                                     │
+                     ┌───────────────┼───────────────┐
+                     ▼               ▼               ▼
+              ┌────────────┐ ┌────────────┐ ┌────────────┐
+              │ type_      │ │ target_    │ │ reconcili- │
+              │ mapping    │ │ profiles   │ │ ation      │
+              │ (类型映射)  │ │ (DDL生成)   │ │ (对数校验)  │
+              └────────────┘ └────────────┘ └────────────┘
+```
+
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/design/de-skills-design-summary.md:293]: Will add language token to the architecture diagram fence.

File: mkdocs.yml [hunk 25]
```diff
@@ -27,6 +27,9 @@ nav:
       - Generate Metrics: subagent/gen_metrics.md
       - Generate Semantic Model: subagent/gen_semantic_model.md
       - Generate SQL Summary: subagent/gen_sql_summary.md
+      - Table Generation: subagent/gen_table.md
+      - ETL Job: subagent/gen_job.md
+      - Migration: subagent/migration.md
```
[the remark this answers is not in the document]
Comment (stukid) [reply mkdocs.yml:32]: Will add zh nav translations for the new entries.

File: tests/integration/tools/test_migration_subagent.py [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+
+    # Cleanup target table using node's own execute_ddl (same SQLAlchemy connection, avoids lock conflicts)
+    try:
+        node.db_func_tool.execute_ddl(sql=f"DROP TABLE IF EXISTS {TARGET_TABLE}", database="greenplum")
+    except Exception:
+        pass
+
+
+@pytest.mark.integration
+class TestGenJobMigrationToolChain:
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/integration/tools/test_migration_subagent.py:112]: Will add `@pytest.mark.nightly` marker in a follow-up.

File: docs/subagent/gen_job.zh.md [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+
+```
+transfer_query_result(
+    source_sql="SELECT * FROM users",
+    source_database="local_duckdb",
+    target_table="public.users_copy",
+    target_database="greenplum",
+    mode="replace"        # replace（TRUNCATE+INSERT）或 append
+)
+```
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/subagent/gen_job.zh.md:147]: Will add language identifiers to fenced blocks.

File: docs/cross-db-migration-design.zh.md [hunk 28]
```diff
@@ -0,0 +1,732 @@
...
+
+## 内置 Subagent 设计：`gen_job`
+
+### 概述
+
+将原有的配置型 `etl` subagent 升级为内置 subagent `gen_job`，参照 `gen_table` 的模式实现。`gen_job` 统一支持单库 ETL 和跨库迁移两种模式。
+
```
[the remark this answers is not in the document]
Comment (stukid) [reply docs/cross-db-migration-design.zh.md]: This file has been removed from the PR (it's a local design doc, not intended for merge).

File: datus/agent/node/migration_agentic_node.py [hunk 29]
```diff
@@ -0,0 +1,314 @@
...
+    - Database tools for schema exploration across multiple databases
+    - DDL execution tool for target table creation
+    - DML execution tool for data writes
+    - Cross-database transfer tool (transfer_query_result)
+    - Filesystem tools for reading/writing SQL artifacts
+    - Ask-user tool for interactive confirmation
+    - Session-based conversation management
+    """
+
+    NODE_NAME = "migration"
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/agent/node/migration_agentic_node.py:44]: Acknowledged. Nightly test coverage for migration in test_node.py will be added in a follow-up PR.

File: datus/agent/node/gen_job_agentic_node.py [hunk 30]
```diff
@@ -0,0 +1,311 @@
...
+    - Database tools for schema exploration
+    - DDL execution tool for table/schema creation
+    - DML execution tool (execute_write) for INSERT/UPDATE/DELETE
+    - Data transfer tool (transfer_query_result) for cross-database migration
+    - Filesystem tools for reading/writing SQL artifacts
+    - Ask-user tool for interactive confirmation
+    - Session-based conversation management
+    """
+
+    NODE_NAME = "gen_job"
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/agent/node/gen_job_agentic_node.py:44]: Acknowledged. Nightly test coverage for gen_job in test_node.py will be added in a follow-up PR.

**Comment by stukid** [comment]
## Response to CodeRabbit Review #3 (Duplicate Comments)

### reconciliation.py: dialect-aware identifier quoting

The `_quote_identifier` currently uses ANSI double quotes. For MySQL/StarRocks compatibility, this should use dialect-aware quoting (backticks). However, this is part of the broader dialect quoting effort tracked in **#563** — the proper fix requires `connector.quote_identifier()` from `datus-db-core`, which is implemented locally but not yet published to PyPI. Will address once the new `datus-db-core` version is released.

### reconciliation.py: composite key distinct_count

Valid point. The per-column `COUNT(DISTINCT col)` misses tuple permutation errors for composite keys. Will fix to use `SELECT COUNT(*) FROM (SELECT DISTINCT key1, key2 FROM table) t` for multi-column keys.

File: datus/tools/func_tool/database.py [hunk 31]
```diff
@@ -1123,22 +1184,371 @@ def execute_ddl(self, sql: str) -> FuncToolResult:
...
+
+            return FuncToolResult(
+                result={
+                    "message": "Write executed successfully",
+                    "sql": normalized_sql,
+                    "sql_type": sql_type.value,
+                    "row_count": row_count,
+                    "database": database or self._default_database,
+                    "dry_run": dry_run,
+                }
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:1348]: Fixed in a06c27eb. When `min_rows`/`max_rows` is requested but connector returns `row_count=None`, now returns `success=0` instead of silently passing.

File: datus/tools/migration/reconciliation.py [hunk 32]
```diff
@@ -0,0 +1,214 @@
...
+# Types considered numeric for aggregate checks
+_NUMERIC_TYPE_PATTERN = re.compile(
+    r"^(INTEGER|INT|INT[248]?|BIGINT|SMALLINT|TINYINT|FLOAT|FLOAT[48]?|DOUBLE|REAL|DECIMAL|NUMERIC|HUGEINT|LARGEINT)",
+    re.IGNORECASE,
+)
+
+# Types considered date/time for min/max checks
+_DATE_TYPE_PATTERN = re.compile(
+    r"^(DATE|TIMESTAMP|DATETIME|TIME)",
+    re.IGNORECASE,
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/migration/reconciliation.py:24]: Fixed in a06c27eb. Added word boundary `\b` to regex patterns to prevent partial matches (e.g., INTERVAL matching INT).

File: datus/tools/migration/reconciliation.py [hunk 33]
```diff
@@ -0,0 +1,214 @@
...
+    return bool(_DATE_TYPE_PATTERN.match(col_type.strip()))
+
+
+def _is_minmax_type(col_type: str) -> bool:
+    return _is_numeric_type(col_type) or _is_date_type(col_type)
+
+
+def _quote_identifier(name: str) -> str:
+    """Quote a SQL identifier with double quotes to handle reserved words and special characters."""
+    return f'"{name}"'
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/migration/reconciliation.py]: Fixed in a06c27eb. `_quote_identifier` now escapes embedded double quotes: `name.replace('"', '""')`.

File: tests/unit_tests/tools/migration/test_reconciliation.py [hunk 34]
```diff
@@ -0,0 +1,304 @@
...
+            key_columns=["id"],
+        )
+        assert len(checks) > 0
+        for check in checks:
+            assert "name" in check
+            assert "source_query" in check
+            assert "target_query" in check
+            assert isinstance(check["source_query"], str)
+            assert len(check["source_query"]) > 0
+
```
[the remark this answers is not in the document]
Comment (stukid) [reply tests/unit_tests/tools/migration/test_reconciliation.py:47]: Fixed in a06c27eb. Added `isinstance(check["target_query"], str)` and `len(check["target_query"]) > 0` assertions.

File: datus/tools/func_tool/database.py [hunk 35]
```diff
@@ -216,8 +216,24 @@ def _get_connector(self, database: Optional[str] = None) -> BaseSqlConnector:
...
+        # Each database in service.databases is its own namespace (see AgentConfig.namespaces).
+        # For cross-database scenarios, use db_name as both namespace and logic_name.
+        try:
+            connector = self._db_manager.get_conn(db_name, db_name)
+        except (KeyError, ValueError) as e:
+            if database:
+                # Caller explicitly requested a specific database — do NOT fall back silently.
+                raise ValueError(
+                    f"Database '{database}' is not configured. Available databases: {', '.join(self._databases)}."
+                ) from e
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:231]: Fixed in a06c27eb. Changed to `DatusException(ErrorCode.COMMON_VALIDATION_FAILED)`.

File: datus/tools/func_tool/database.py [hunk 36]
```diff
@@ -749,7 +776,24 @@ def list_databases(self, catalog: Optional[str] = "", include_sys: Optional[bool
...
+                try:
+                    self._get_connector(name)
+                except Exception as e:
+                    available = False
+                    error_msg = str(e)
+                entry = {"name": name, "type": db_type, "available": available}
+                if not available:
+                    entry["error"] = error_msg
+                db_list.append(entry)
+            return FuncToolResult(success=1, result=db_list)
```
[the remark this answers is not in the document]
Comment (stukid) [reply datus/tools/func_tool/database.py:800]: Fixed in a06c27eb. Added `_database_matches_scope()` filter to multi-connector `list_databases()` branch.
